gpu/octree/src/cuda/approx_nsearch.cu

```diff
@@ -106,38 +106,112 @@ namespace pcl { namespace device { namespace appnearest_search
 	private:
 
 		__device__ __forceinline__ int findNode()
-		{                                             
+		{
 			float3 minp = batch.octree.minp;
 			float3 maxp = batch.octree.maxp;
 
-			if(query.x < minp.x || query.y < minp.y ||  query.z < minp.z)
-				return 0;
-
-			if(query.x > maxp.x || query.y > maxp.y ||  query.z > maxp.z)
-				return 0;
-
 			int node_idx = 0;
 			int code = CalcMorton(minp, maxp)(query);
 			int level = 0;
 
+			bool centroid_traversal = false;
+			int mask_pos;
+			int x, y, z;
+
 			for(;;)
 			{
-				int mask_pos = 1 << Morton::extractLevelCode(code, level);
-
 				int node = batch.octree.nodes[node_idx];
 				int mask = node & 0xFF;
 
+				float3 query_point;
+				query_point.x = query.x;
+				query_point.y = query.y;
+				query_point.z = query.z;
+
 				if(__popc(mask) == 0)  // leaf
-					return node_idx; 
+				{
+					//printf ("node x %d\n", node_idx);
+					return node_idx;
+				}
+
+				if (!centroid_traversal)    // no empty voxel encountered yet, performing morton code based traversal
+				{
+					mask_pos = 1 << Morton::extractLevelCode(code, level);
 
-				if ( (mask & mask_pos) == 0) // no child
-					return node_idx;                         
+					if ( (mask & mask_pos) == 0) // no child
+					{
+
+						//find current cell
+						Morton::decomposeCode(code, x, y, z);
+
+						x >>= (Morton::levels - level);
+						y >>= (Morton::levels - level);
+						z >>= (Morton::levels - level);
+
+						centroid_traversal = true;  //switch to nearest-centroid based traversal
+						mask_pos = nearestVoxelTraversal(query_point, level, mask, minp, maxp, x, y, z);
+					}
+				}
+
+				else
+					mask_pos = nearestVoxelTraversal(query_point, level, mask, minp, maxp, x, y, z);
 
 				node_idx = (node >> 8) + __popc(mask & (mask_pos - 1));
 				++level;
 			}
 		};
 
+		__device__ int nearestVoxelTraversal(float3 query, int level, int mask, float3 minp, float3 maxp, int& x, int& y, int& z)
```

Comment:
this one is also the same as the other method in octree_host. we should not keep  separate functions which do exactly the same thing.

Reply:
Same comment with respect to reviewing this.